Editing on Basic Data Types (#1827)
ocaml/ocaml.org: data/tutorials/language/0it_01_basic_datatypes.md

## Introduction

In OCaml, there are no type checks at runtime, and values don't change type unless explicitly converted. This is what being statically- and strongly-typed means. This allows convenient and safe processing of structured data. 

> This tutorial introduces basic data types in OCaml. _Variants_ and _products_ represent data in OCaml. These correspond to [algebraic data types](https://en.wikipedia.org/wiki/Algebraic_data_type). At this level, a [nominal](https://en.wikipedia.org/wiki/Nominal_type_system) type-checking algorithm is used. Historically, this is OCaml's first type system, as it comes from the [ML](https://en.wikipedia.org/wiki/ML_(programming_language)) programming language, OCaml's ancestor. Although OCaml has other type systems, this document focuses on data typed using this algorithm.

Author: ```suggestion

```

<!-- It is also an expression-oriented language: everything is a value and every value has a type. Functions and types are the two foundational principles of OCaml. The OCaml type system is highly expressive, providing many advanced constructs while being easy to use and unobtrusive. Thanks to type inference, programs can be written without type annotations, except for documentation purposes and a few corner cases. The basic types and the type combination operations enable a vast range of possibilities.

> Type annotation means that it's necessary to write the value type to tell the compiler what to do, as opposed to having the compiler infer the type.
-->

<!--ORIGINAL
You will learn about atomic types, such as integers and Booleans; predefined compound types, like strings and lists; and user-defined types, namely variants and records. The goal is to teach skills on how to handle data from predefined, variant, and record types. You will also be introduced to the important concept of pattern matching on those types.
-->

<!--More scannable-->

> This course covers various types: 
> - Atomic types, such as integers and Booleans
> - Predefined compound types, like strings and lists

Author: ```suggestion
This course covers: atomic types, such as integers and Booleans; predefined compound types, like strings and lists; and user-defined types, namely variants and records.
```

> - User-defined types, namely variants and records

Author: ```suggestion
```

> This course covers various types: 
> - Atomic types, such as integers and Booleans
> - Predefined compound types, like strings and lists
> - User-defined types, namely variants and records
> 
> The goal is to teach skills on how to handle data from predefined, variant, and record types. Additionally, you will be introduced to the important concept of pattern matching on those types.

Author: ```suggestion
The goal is to teach skills on how to handle:
- Atomic types, such as integers and Booleans
- Predefined compound types, like strings and lists
- User-defined types, namely variants and records
 
Additionally, you will be introduced to the important concept of pattern matching on those types.
```

> The goal is to teach skills on how to handle data from predefined, variant, and record types. Additionally, you will be introduced to the important concept of pattern matching on those types.

Author: ```suggestion
Additionally, you will be introduced to the important concept of pattern matching on those types.
```

<!--End edit-->

> Before proceeding, it's necessary to have completed the [Get Started](https://ocaml.org/docs/get-started) series of tutorials as well as [Functions and Values](/docs/values-and-functions). As in previous tutorials, expressions after `#` and ending with `;;` are for the toplevel, like UTop.

Author: ```suggestion
**Prerequisites**: Before proceeding, it's necessary to have completed the [Get Started](https://ocaml.org/docs/get-started) series of tutorials as well as [Functions and Values](/docs/values-and-functions). As in previous tutorials, expressions after `#` and ending with `;;` are for the toplevel, like UTop.
```

> Before proceeding, it's necessary to have completed the [Get Started](https://ocaml.org/docs/get-started) series of tutorials as well as [Functions and Values](/docs/values-and-functions). As in previous tutorials, expressions after `#` and ending with `;;` are for the toplevel, like UTop.

Author: ```suggestion
**Prerequisites**: Before proceeding, it's necessary to have completed the [Get Started](https://ocaml.org/docs/get-started) series of tutorials as well as [Functions and Values](/docs/values-and-functions). As in previous tutorials, expressions after `#` and ending with `;;` are for the toplevel, like UTop.
```

#### Byte Sequences

<!--CR
Moved intro pp to after example, as it contains the definition. Perhaps a short intro sentence here for context?
--> 

```ocaml
# String.to_bytes "hello";;
- : bytes = Bytes.of_string "hello"
```

> Byte sequences are finite and fixed-sized. Each individual byte is represented by a `char` value. Byte sequences are mutable, meaning they can't be extended or shortened, but each component byte may be updated. Essentially, a byte sequence (type `bytes`) is a mutable string that can't be printed. There is no way to write `bytes` literally, so they must be produced by a function.

Reviewer: ```suggestion
Like strings, byte sequences are finite and fixed-sized. Each individual byte is represented by a `char` value. Like arrays, byte sequences are mutable, meaning they can't be extended or shortened, but each component byte may be updated. Essentially, a byte sequence (type `bytes`) is a mutable string that can't be printed. There is no way to write `bytes` literally, so they must be produced by a function.
```

> Operations on `bytes` values are provided by the [`Stdlib`](/api/Stdlib.html) and the [`Bytes`](/api/Bytes.html) modules. Only the function `Bytes.get` allows direct access to the characters contained in a byte sequence. There is no direct access operator on byte sequences.

Reviewer: ```suggestion
Operations on `bytes` values are provided by the [`Stdlib`](/api/Stdlib.html) and the [`Bytes`](/api/Bytes.html) modules. Only the function `Bytes.get` allows direct access to the characters contained in a byte sequence. Unlike arrays, there is no direct access operator on byte sequences. 
```

#### Byte Sequences

<!--CR
Moved intro pp to after example, as it contains the definition. Perhaps a short intro sentence here for context?
--> 

```ocaml
# String.to_bytes "hello";;
- : bytes = Bytes.of_string "hello"
```
Like strings, byte sequences are finite and fixed-sized. Each individual byte is represented by a `char` value. Like arrays, byte sequences are mutable, meaning they can't be extended or shortened, but each component byte may be updated. Essentially, a byte sequence (type `bytes`) is a mutable string that can't be printed. There is no way to write `bytes` literally, so they must be produced by a function.

Operations on `bytes` values are provided by the [`Stdlib`](/api/Stdlib.html) and the [`Bytes`](/api/Bytes.html) modules. Only the function `Bytes.get` allows direct access to the characters contained in a byte sequence. Unlike arrays, there is no direct access operator on byte sequences. 

> 

Author: ```suggestion

The memory representation of `bytes` is four times more compact that `char array`.
```
